torchtext/csrc/vocab.cpp

```diff
@@ -0,0 +1,117 @@
+#include <stdexcept>
+#include <string>
+#include <torch/script.h>
+
+namespace torchtext {
+namespace {
+
+struct Vocab : torch::CustomClassHolder {
+private:
+  int64_t unk_index_;
+  std::unordered_map<std::string, int64_t> stoi_;
+
+public:
+  // stoi_, and unordered_map holds the serialized params passed in
+  // during initialization. We need this because we need to be able to serialize
+  // the model so that we can save the scripted object. Pickle will get the
+  // serialized model from these members, thus they needs to be public.
+  std::vector<std::string> itos_;
+  std::string unk_token_;
+
+  explicit Vocab(const std::vector<std::string> &tokens,
+                 const std::string &unk_token)
+      : itos_(tokens), unk_token_(unk_token) {
+    stoi_.reserve(tokens.size());
+    for (std::size_t i = 0; i < tokens.size(); i++) {
+      // tokens should not have any duplicates
+      if (stoi_.find(tokens[i]) != stoi_.end()) {
+        throw std::runtime_error("Duplicate token found in tokens list: " +
+                                 tokens[i]);
+      }
+      stoi_[tokens[i]] = i;
+    }
+    unk_index_ = stoi_[unk_token];
+  }
+
+  int64_t __len__() const { return stoi_.size(); }
+
+  int64_t __getitem__(const std::string &token) const {
+    const auto &item = stoi_.find(token);
+    if (item != stoi_.end()) {
+      return item->second;
+    }
+    return unk_index_;
+  }
+
+  void addToken(const std::string &token) {
```

Comment:
As my comments above, we need more discussions for this func.

Reply:
> Are we going to support a similar func, like [lookup_indices_1d](https://github.com/facebookresearch/pytext/blob/2c6a2b602cbb9e784c5cb4abdb21d26983f12d83/pytext/torchscript/vocab.py#L35)?
> 
> Do we want to have a func for [lookup_word](https://github.com/facebookresearch/pytext/blob/2c6a2b602cbb9e784c5cb4abdb21d26983f12d83/pytext/torchscript/vocab.py#L102)?

Let's discuss this further. We can probably add this in follow up PRs.